File: python/packages/mcpfy-pulse/tests/fixtures/shutdown_child.py
"""Spawned as a real subprocess by test_shutdown_flush.py. Simulates a server
process getting killed directly (e.g. by a process manager, `kill`, or a
container orchestrator) rather than via a client's graceful stdin-close — the
one shutdown path the pre-existing anyio.EndOfStream-triggered flush in
streams.py cannot cover on its own, since no EOF ever happens here."""

import anyio
from mcp.shared.message import SessionMessage
from mcp.types import JSONRPCMessage, JSONRPCRequest, JSONRPCResponse

from mcpfy_pulse import TelemetryOptions, with_mcpfy_telemetry


async def main() -> None:
    read_send, read_recv = anyio.create_memory_object_stream(10)
    write_send, write_recv = anyio.create_memory_object_stream(10)

    wrapped_read, wrapped_write = with_mcpfy_telemetry(
        read_recv, write_send, TelemetryOptions(server_name="shutdown-test-server")
    )

    # Simulate a real tools/call round trip so one real event ends up queued.
    await read_send.send(
        SessionMessage(
            JSONRPCMessage(
                JSONRPCRequest(jsonrpc="2.0", id=1, method="tools/call", params={"name": "shutdown-test-tool"})
            )
        )
    )
    await wrapped_read.receive()
    await wrapped_write.send(
        SessionMessage(JSONRPCMessage(JSONRPCResponse(jsonrpc="2.0", id=1, result={"content": []})))
    )

    # Keep the process alive so the test controls exactly when it dies, via a
    # real SIGTERM sent directly to this process.
    await anyio.sleep_forever()


if __name__ == "__main__":
    anyio.run(main)

File: typescript/packages/mcpfy-pulse/tests/fixtures/shutdown-child.ts
import { withMcpfyTelemetry } from "../../src/wrap-transport.js";
import type { MinimalTransport } from "../../src/types.js";

let onmessageHandler: ((message: any, extra?: any) => void) | undefined;

const fakeTransport: MinimalTransport = {
  start: async () => {},
  close: async () => {},
  send: async (_message: any) => {},
  get onmessage() {
    return onmessageHandler;
  },
  set onmessage(handler) {
    onmessageHandler = handler;
  },
  onclose: undefined,
  onerror: undefined,
};

const wrapped = withMcpfyTelemetry(fakeTransport, { serverName: "shutdown-test-server" });
wrapped.onmessage = () => {};

// Simulate a real tools/call round trip — exactly what a real client/server
// exchange produces — so one real event ends up sitting in the batcher's queue.
onmessageHandler?.({ jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "shutdown-test-tool" } });
await wrapped.send({ jsonrpc: "2.0", id: 1, result: { content: [] } });

// Keep the process alive (unref'd timers, like the batcher's own flush timer,
// would let it exit immediately otherwise) so the test controls exactly when
// this process dies, via a real SIGTERM — mirroring a server idling on stdio.
setInterval(() => {}, 1000);
